packages/tpetra/core/src/Tpetra_Import_Util2.hpp

```diff
@@ -946,6 +952,287 @@ lowCommunicationMakeColMapAndReindex (const Teuchos::ArrayView<const size_t> &ro
+  auto rowptr_view = Details::create_mirror_view_from_raw_host_array(DT(), rowptr.getRawPtr(), rowptr.size(), true, "rowptr");
+  auto colind_GID_view = Details::create_mirror_view_from_raw_host_array(DT(), colind_GID.getRawPtr(), colind_GID.size(), true, "colind_GID");
+  auto owningPIDs_view = Details::create_mirror_view_from_raw_host_array(DT(), owningPIDs.getRawPtr(), owningPIDs.size(), true, "owningPIDs");
+
+  typename decltype(colind_LID_view)::HostMirror colind_LID_host(colind_LID.getRawPtr(), colind_LID.size());
+  typename decltype(colind_GID_view)::HostMirror colind_GID_host(colind_GID.getRawPtr(), colind_GID.size());
+
+  Kokkos::deep_copy(colind_LID_view, colind_LID_host);
+  Kokkos::deep_copy(colind_GID_view, colind_GID_host);
+
+  // The domainMap is an RCP because there is a shortcut for a
+  // (common) special case to return the columnMap = domainMap.
+  const map_type& domainMap = *domainMapRCP;
+
+  int myPID = domainMap.getComm()->getRank ();
+
+
+  // We use two hash tables, one for local GIDs and one for remote GIDs
+  // FIXME (mfh 11 Feb 2015) Tpetra::Details::HashTable can hold at
+  // most INT_MAX entries, but it's possible to have more rows than
+  // that (if size_t is 64 bits and int is 32 bits).
+  Kokkos::UnorderedMap<LO, bool, execution_space> LocalGIDs_view_map(colind_LID.size());
+  Kokkos::UnorderedMap<GO, LO, execution_space> RemoteGIDs_view_map(colind_LID.size());
+  
+  const size_t numMyRows = rowptr.size () - 1;
+  local_map_type domainMap_local = domainMap.getLocalMap();
+  
+  const size_t numDomainElements = domainMap.getLocalNumElements ();
+  Kokkos::View<bool*> LocalGIDs_view("LocalGIDs", numDomainElements);
+  auto LocalGIDs_host = Kokkos::create_mirror_view(LocalGIDs_view);
+  
+  size_t NumLocalColGIDs = 0;
+  
+  // Scan all column indices and sort into two groups:
+  // Local:  those whose GID matches a GID of the domain map on this processor and
+  // Remote: All others.
+  // Kokkos::Parallel_reduce sums up NumLocalColGIDs, while we use the size of the Remote GIDs map to find NumRemoteColGIDs
+  Kokkos::parallel_reduce(team_policy(numMyRows, Kokkos::AUTO), KOKKOS_LAMBDA(const typename team_policy::member_type &member, size_t &update) { 
+    const int i = member.league_rank();
+    size_t NumLocalColGIDs_temp = 0;
+    size_t rowptr_start = rowptr_view[i];
+    size_t rowptr_end = rowptr_view[i+1];
+    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(member, rowptr_start, rowptr_end), [&](const size_t j, size_t &innerUpdate) {
+      const GO GID = colind_GID_view[j];
+      // Check if GID matches a row GID in local domain map
+      const LO LID = domainMap_local.getLocalElement (GID);
+      if (LID != -1) {
+        auto outcome = LocalGIDs_view_map.insert(LID);
+        // Fresh insert
+        if(outcome.success()) {
+          LocalGIDs_view[LID] = true;
+          Kokkos::atomic_add(&innerUpdate, 1);
```

Comment:
You don't have to use atomics to update partial reductions - Kokkos promises that no other thread gets the same reference at the same time